src/rust/cryptography-x509-validation/src/policy/mod.rs

```diff
@@ -234,15 +249,340 @@ impl<'a, B: CryptoOps> Policy<'a, B> {
+        self.permits_basic(cert)?;
+
+        // 5280 4.1.2.6: Subject
+        // CA certificates MUST have a subject populated with a non-empty distinguished name.
+        // No check required here: `permits_basic` checks that the issuer is non-empty
+        // and `ChainBuilder::potential_issuers` enforces subject/issuer matching,
+        // meaning that an CA with an empty subject cannot occur in a built chain.
+
+        // NOTE: This conceptually belongs in `valid_issuer`, but is easier
+        // to test here. It's also conceptually an extension policy, but
+        // requires a bit of extra external state (`current_depth`) that isn't
+        // presently convenient to push into that layer.
+        if let Some(bc) = extensions.get_extension(&BASIC_CONSTRAINTS_OID) {
+            let bc: BasicConstraints = bc.value()?;
+
+            if bc
+                .path_length
+                .map_or(false, |len| current_depth as u64 > len)
+            {
+                return Err(ValidationError::Other(
+                    "path length constraint violated".to_string(),
+                ))?;
+            }
+        }
+
+        for ext_policy in self.ca_extension_policies.iter() {
+            ext_policy.permits(self, cert, extensions)?;
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether the given EE certificate is compatible with this policy.
+    pub(crate) fn permits_ee(
+        &self,
+        cert: &Certificate<'_>,
+        extensions: &Extensions<'_>,
+    ) -> Result<(), ValidationError> {
+        self.permits_basic(cert)?;
+
+        for ext_policy in self.ee_extension_policies.iter() {
+            ext_policy.permits(self, cert, extensions)?;
+        }
+
+        Ok(())
+    }
+
+    /// Checks whether `issuer` is a valid issuing CA for `child` at a
+    /// path-building depth of `current_depth`.
+    ///
+    /// This checks that `issuer` is permitted under this policy and that
+    /// it was used to sign for `child`.
+    ///
+    /// On success, this function returns the new path-building depth. This
+    /// may or may not be a higher number than the original depth, depending
+    /// on the kind of validation performed (e.g., whether the issuer was
+    /// self-issued).
+    pub(crate) fn valid_issuer(
+        &self,
+        issuer: &Certificate<'_>,
+        child: &Certificate<'_>,
+        current_depth: u8,
+        issuer_extensions: &Extensions<'_>,
+    ) -> Result<u8, ValidationError> {
+        // The issuer needs to be a valid CA at the current depth.
+        self.permits_ca(issuer, current_depth, issuer_extensions)?;
+
+        // CA/B 7.1.3.1 SubjectPublicKeyInfo
+        if !self
+            .permitted_public_key_algorithms
+            .contains(&child.tbs_cert.spki.algorithm)
+        {
+            return Err(ValidationError::Other(format!(
+                "Forbidden public key algorithm: {:?}",
+                &child.tbs_cert.spki.algorithm
+            )));
+        }
```

Comment:
This is because of the snafu we ran into with old `sha1WithRsaEncryption` root CAs: putting this check in `permits_basic` means that we'll end up running it on *every* cert in the chain, including ones whose signatures we don't actually want to verify (since they're in the trust root and therefore aren't required to have CABF-conforming self-sigs).

Let me know if you think I should drop a comment to that effect here; there might also be another place to put these that I've missed.